lldb/source/Core/Debugger.cpp

```diff
@@ -743,10 +743,11 @@ DebuggerSP Debugger::CreateInstance(lldb::LogOutputCallback log_callback,
 }
 
 void Debugger::HandleDestroyCallback() {
-  if (m_destroy_callback) {
-    m_destroy_callback(GetID(), m_destroy_callback_baton);
-    m_destroy_callback = nullptr;
+  const lldb::user_id_t user_id = GetID();
```

Comment:
A destroy callback could delete another destroy callback while running.  You need to account for that.

Reply:
I don't understand.

Were you referring to the `foreach` loop below and saying that since the container can be modified during the loop, the looping needs to be changed (e.g. to use the iterator instead of the `foreach` syntax)?

If you did meant to talk about the `user_id` then I don't think I understand how the `user_id` is related, because it seems it should be constant throughout the `HandleDestroyCallback` execution.

Reply:
I meant the former.  You could have added two Destroy callbacks, but then during the execution of the first one, the code figures out that it has already done the job the second one was intended to do, so the first Destroy callback removed the second one on the way out to keep the work from getting done twice.

Reply:
Added double while-loop in `HandleDestroyCallback` to handle callbacks which can be added/removed during destroy. Added test `test_HandleDestroyCallback` to validate a simple example case.